client/src/extension.ts

```diff
@@ -214,21 +214,49 @@ export async function activate (context: vscode.ExtensionContext): Promise<void>
   }))
 
-  // Check if the document that was just closed was the last one for a recipe
-  context.subscriptions.push(vscode.workspace.onDidCloseTextDocument((document: vscode.TextDocument) => {
-    const ext = ['.bb', '.bbappend', '.inc']
-    const { fsPath } = document.uri
-    if (ext.includes(path.extname(fsPath))) {
-      const recipeName = extractRecipeName(fsPath)
-      const recipeFile = vscode.window.visibleTextEditors.find((editor) => {
-        return ext.includes(path.extname(editor.document.uri.fsPath)) && extractRecipeName(editor.document.uri.fsPath) === recipeName
-      })
-      if (recipeFile === undefined) {
-        logger.debug(`No files related to the recipe ${recipeName}, sending notification to remove scan results`)
-        void client.sendNotification('bitbake/removeScanResult', { recipeName })
+  context.subscriptions.push(
+    // Check if the document that was just closed was the last one for a recipe
+    vscode.workspace.onDidCloseTextDocument((document: vscode.TextDocument) => {
+      const ext = ['.bb', '.bbappend', '.inc']
+      const { fsPath } = document.uri
+      if (ext.includes(path.extname(fsPath))) {
+        const recipeName = extractRecipeName(fsPath)
+        const recipeFile = vscode.window.visibleTextEditors.find((editor) => {
+          return ext.includes(path.extname(editor.document.uri.fsPath)) && extractRecipeName(editor.document.uri.fsPath) === recipeName
+        })
+        if (recipeFile === undefined) {
+          logger.debug(`No files related to the recipe ${recipeName}, sending notification to remove scan results`)
+          void client.sendNotification('bitbake/removeScanResult', { recipeName })
+        }
       }
-    }
-  }))
+    }),
+    // Run different bitbake commands based on the saved document type
+    vscode.workspace.onDidSaveTextDocument(async (document: vscode.TextDocument) => {
+      if (document.languageId !== 'bitbake') {
+        // Embedded files also trigger this event, but we don't do anything with them
+        return
+      }
+      const parseOnSave = vscode.workspace.getConfiguration('bitbake').get('parseOnSave')
+      if (parseOnSave !== true) {
+        return
+      }
+      const exts = ['.bb', '.bbappend', '.inc']
+      const { fsPath } = document.uri
+
+      if (exts.includes(path.extname(fsPath))) {
+        const foundRecipe = bitBakeProjectScanner.scanResult._recipes.find((recipe) => recipe.name === extractRecipeName(fsPath))
+        if (foundRecipe !== undefined) {
+          logger.debug(`[onDidSave] Running 'bitbake -e' against the saved recipe: ${foundRecipe.name}`)
+          // Note that it pends only one scan at a time. See client/src/driver/BitbakeRecipeScanner.ts.
+          // Saving more than 2 files at the same time could cause the server to miss some of the scans.
+          await vscode.commands.executeCommand('bitbake.scan-recipe-env', document.uri)
+        }
+        return

```

Comment:
I think this return statement should be moved the line above. Otherwise, thereis a possibility that we don't launch the generic "parse-recipes" task.

Reply:
Yes, when the saved file doesn't match any recipe we scanned, it won't trigger anything. I will move it below the command (line 252). So, in what scenario does our project scanner fail to scan a file with the correct recipe extension?

Reply:
This could happen but it's not an error. Bitbake can mask recipes so that a .inc/.bb/.bbappend file can be present but excluded from the build system. Another way this can happend with a file that is actually part of the build system is a .inc with a name not related with the current recipe. Example: disable-cache.inc which would be included by some-recipe.bb.

Reply:
In those cases we still want to run the global parsing as a fallback.